_posts/2014-12-28-front-end-streams.md

```diff
@@ -0,0 +1,271 @@
+  returned stream.
+* Whenever our stream is written to, it resets the context of the Ractive
+  instance.
+
+Now, let's build our theoretical widget! We can assume that it will be handed a
+DOM element to render itself into.
+
+Let's start with the template
+
+```html
+<div class="card-input">
+  <input type="text" name="full-name"/>
+  <input type="text" name="card-number"/>
+</div>
+<div class="card-output">
+  <div class="card-template">
+    <span class="card-number">{{ cardNumber }}</span>
+    <span class="card-name">{{ fullName }}</span>
+  </div>
+</div>
+```
+
+Simple enough, now for the JavaScript:
+
+```javascript
+var events = require('dom-delegation-stream')
+  , values = require('dom-value-stream')
+  , objectState = require('objectstate')
+
+var ractiveStream = require('./ractive-stream')
+  , cardTemplate = require('./template.ract') // check out the ractify transform
+  , stripNonDigits = require('./strip-non-digits')
+  , addDashes = require('./add-dashes')
+
+module.exports = cardWidget
+
+function cardWidget(el) {
+  var ractive = ractiveStream(el, cardTemplate)
+    , state = objectState()
+
+  // DOM events -> values, assign `name` to the values stream
+  // we select on the `name` attribute, because it is not typically subject to
+  // change as frequently as class or ID.
+  var name = events(el, 'input', '[name=full-name]')
+    .pipe(values())
+
+  // DOM events -> values -> strip non-digits -> add dashes, assign `cardNumber`
+  // to the dashes stream
+  // " 123 45678,91011a12" -> "123456789101112" -> "1234-5678-9101-112"
+  var cardNumber = events(el, 'input', '[name=card-number]')
+    .pipe(values())
+    .pipe(stripNonDigits())
+    .pipe(addDashes())
+
+  state.listen(name, 'fullName')
+    .listen(cardNumber, 'cardNumber')
+
+  state.pipe(ractive)
+}
+```
+
+Let's break it down:
+
+* We require all of the modules we will be using, including a couple of
+  theoretical but trivially-implemented modules that export a function that
+  returns a stream:
+  - "strip-non-digits" just removes all non-digit characters from a string.
+  - "add-dashes" inserts dashes into a string in a pattern like you might
+    expect on an ID card.
+* We declare our export, which is the function that constructs this "widget"
+* Our function takes an element, and immediately creates a ractiveStream with
+  that element and our template.
+* We initialize an ObjectState to serve as our source of context.
+* We create two pipelines:
+  - They both start as events -> values, effectively turning input keystrokes
+    into the value of the element they were triggered from.
+  - The `cardNumber` stream is further piped into our "add-dashes" module for
+    formatting.
+* Our objectState is told to listen to these streams, and consider them as keys
+  "fullName" and "cardNumber" respectively.
+* We pipe our objectState directly into our ractiveStream to be used as context.
```

Comment:
I'm trying to think of another way to present this information, but this as bullet points caused me to do a bunch of scrolling up and down to look at the example code and back at the point.  I almost imagine a screen capture of the example code with these bullet points each as a comment for the code it refers to.